### DISCUSSION TITLE
Using mingw64 Cross compilation  to compile strongswan in ubuntu environment failed, when include <openssl/ax509.h>

### DISCUSSION THREAD

**Discussion by xtsheng123** [strongswan/strongswan#677] [General]
1. environment :     ubuntu  mingw64  strongswan-5.6.3  openssl-1.1.1d
2. Compiling OpenSSL:    `CROSS_COMPILE="x86_64-w64-mingw32-" ./Configure mingw64 no-asm shared --prefix=/opt/mingw64`
3. Compiling strongswan : 
`CFLAGS="-g -O2 -Wall -Wno-pointer-sign -Wno-format-security \
        -Wno-format -mno-ms-bitfields \
        -I/root/project/strongswan/install_file/include" \
LDFLAGS="-L/root/project/strongswan/install_file/lib" \
./configure --disable-defaults --enable-monolithic --enable-static \
--enable-svc --enable-ikev2 --enable-ikev1  \
--enable-nonce --enable-pem --enable-pkcs1 \
--enable-x509 --enable-openssl --enable-socket-win \
--enable-kernel-wfp --enable-kernel-iph --enable-pubkey \
--enable-swanctl --with-swanctldir=swanctl \
--with-strongswan-conf=strongswan.conf --host=x86_64-w64-mingw32 --prefix=/root/project/strongswan/install_file`
4. error information:
`libtool: compile:  x86_64-w64-mingw32-gcc -DHAVE_CONFIG_H -I. -I../../../.. -I../../../../src/libstrongswan -DFIPS_MODE=0 -I/root/project/strongswan/install_openssl-1.1.1d/include -g -O2 -Wall -Wno-pointer-sign -Wno-format-security -Wno-format -mno-ms-bitfields -I/root/project/strongswan/install_openssl-1.1.1d/include -include /root/project/strongswan/strongswan-5.6.3/config.h -MT openssl_ec_private_key.lo -MD -MP -MF .deps/openssl_ec_private_key.Tpo -c openssl_ec_private_key.c  -DDLL_EXPORT -DPIC -o .libs/openssl_ec_private_key.o
In file included from /usr/share/mingw-w64/include/windows.h:95:0,
                 from /usr/share/mingw-w64/include/winsock2.h:23,
                 from ../../../../src/libstrongswan/utils/compat/windows.h:24,
                 from ../../../../src/libstrongswan/utils/utils.h:39,
                 from ../../../../src/libstrongswan/library.h:101,
                 from ../../../../src/libstrongswan/credentials/builder.h:41,
                 from openssl_ec_private_key.h:26,
                 from openssl_ec_private_key.c:21:
/root/project/strongswan/install_openssl-1.1.1d/include/openssl_/x509.h:88:34: error: expected ‘)’ before numeric constant_
 typedef STACK_OF(X509_EXTENSION) X509_EXTENSIONS;
                                  ^
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/pkcs7.h:47:3: error: expected ‘)’ before numeric constant
 } PKCS7_SIGNER_INFO;
   ^
In file included from /root/project/strongswan/install_openssl-1.1.1d/include/openssl/crypto.h:23:0,
                 from /root/project/strongswan/install_openssl-1.1.1d/include/openssl/bio.h:20,
                 from /root/project/strongswan/install_openssl-1.1.1d/include/openssl/evp.h:16,
                 from openssl_ec_private_key.h:24,
                 from openssl_ec_private_key.c:21:
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:20:32: **_error: pasting "stack_st_" and_** "(" does not give a valid preprocessing token
 # define STACK_OF(type) struct stack_st_##type
                                ^
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:23:5: note: in expansion of macro ‘STACK_OF’
     STACK_OF(t1); \
     ^~~~~~~~
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:130:29: note: in expansion of macro ‘SKM_DEFINE_STACK_OF’
 # define DEFINE_STACK_OF(t) SKM_DEFINE_STACK_OF(t, t, t)
                             ^~~~~~~~~~~~~~~~~~~
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/pkcs7.h:49:1: note: in expansion of macro ‘DEFINE_STACK_OF’
 DEFINE_STACK_OF(PKCS7_SIGNER_INFO)
 ^
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/pkcs7.h:49:17: error: expected ‘)’ before numeric constant
 DEFINE_STACK_OF(PKCS7_SIGNER_INFO)
                 ^
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:20:43: note: in definition of macro ‘STACK_OF’
 # define STACK_OF(type) struct stack_st_##type
                                           ^~~~
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:130:29: note: in expansion of macro ‘SKM_DEFINE_STACK_OF’
 # define DEFINE_STACK_OF(t) SKM_DEFINE_STACK_OF(t, t, t)
                             ^~~~~~~~~~~~~~~~~~~
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/pkcs7.h:49:1: note: in expansion of macro ‘DEFINE_STACK_OF’
 DEFINE_STACK_OF(PKCS7_SIGNER_INFO)
 ^
/root/project/strongswan/install_openssl-1.1.1d/include/openssl/safestack.h:24:19: error: pasting "sk_" and "(" does not give a valid preprocessing token
     typedef int (*sk_##t1##_compfunc)(const t3 * const *a, const t3 *const *b); \`

**Comment by tobiasbrunner** [comment]
You can try if 55df5e979718a80015e858ca32bc2c92d08b39cf helps, but I'd recommend you use a newer version of strongSwan.

**Reply by xtsheng123** [reply]
Think you very much.  I have tried your method 55df5e9 , this problem is solved, but there are other problems. I am trying to solve new problems.  Thank you again!!
